File: Controller.py
import pygame
pygame.init()
class Controller():
    def __init__(self, CPU):
        self.CPU = CPU
        self.controls1 = 0xFF
        self.controls2 = 0xFF
        
    def Update(self):
        controls1 = 0b11111111
        UP =    0b11111110
        DOWN =  0b11111101
        LEFT =  0b11111011
        RIGHT = 0b11110111
        POWER = 0b11101111
        MENU =  0b11011111

        keys=pygame.key.get_pressed()
        if keys[pygame.K_RIGHT]:
            controls1 &= RIGHT
        if keys[pygame.K_LEFT]:
            controls1 &= LEFT
        if keys[pygame.K_DOWN]:
            controls1 &= DOWN
        if keys[pygame.K_UP]:
            controls1 &= UP
        if keys[pygame.K_RCTRL]:
            controls1 &= MENU
        if keys[pygame.K_p]:
            controls1 &= POWER

        
        controls2 = 0b11111111
        SCREEN_TOP_LEFT =      0b11111110
        SCREEN_TOP_RIGHT =     0b11111101
        SCREEN_BOTTOM_LEFT =   0b11111011
        SCREEN_BOTTOM_RIGHTT = 0b11110111
        
        if keys[pygame.K_q]:
            controls2 &= SCREEN_TOP_LEFT
        if keys[pygame.K_w]:
            controls2 &= SCREEN_TOP_RIGHT
        if keys[pygame.K_s]:
            controls2 &= SCREEN_BOTTOM_LEFT
        if keys[pygame.K_a]:
            controls2 &= SCREEN_BOTTOM_RIGHTT
        


        if (controls1 != self.controls1) or (controls2 != self.controls2):
            self.CPU.PTInterrupt.Trigger()

        self.controls1 = controls1
        self.controls2 = controls2

        self.CPU.MMU.StoreMemory(0, self.controls1)
        self.CPU.MMU.StoreMemory(1, self.controls2)

        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                pygame.quit()
                exit()

File: Controller.cpp
#include "Controller.h"
#include "CPU.h"
#include "MMU.h"
#include "Interrupt.h"

#define UP                   0b1111111111111110
#define DOWN                 0b1111111111111101
#define LEFT                 0b1111111111111011
#define RIGHT                0b1111111111110111
#define POWER                0b1111111111101111
#define MENU                 0b1111111111011111
#define UPSIDE_UP            0b1111111110111111
#define UPSIDE_DOWN          0b1111111101111111
#define SCREEN_TOP_LEFT      0b1111111011111111
#define SCREEN_TOP_RIGHT     0b1111110111111111
#define SCREEN_BOTTOM_LEFT   0b1111101111111111
#define SCREEN_BOTTOM_RIGHT  0b1111011111111111
#define ACTION               0b1110111111111111
#define MUTE                 0b1101111111111111


Controller::Controller(CPU* cpu){
    this->cpu = cpu;
    this->controls = 0xFFFF;
}

void Controller::Update(){
    const uint8_t *state = SDL_GetKeyboardState(NULL);

    unsigned short keys = 0b1111111111111111;

    if (state[SDL_SCANCODE_RIGHT]){
        keys &= RIGHT;
    }
    if (state[SDL_SCANCODE_LEFT]){
        keys &= LEFT;
    }
    if (state[SDL_SCANCODE_DOWN]){
        keys &= DOWN;
    }
    if (state[SDL_SCANCODE_UP]){
        keys &= UP;
    }
    if (state[SDL_SCANCODE_RCTRL]){
        keys &= MENU;
    }
    if (state[SDL_SCANCODE_P]){
        keys &= POWER;
    }
    if (state[SDL_SCANCODE_A]){
        keys &= ACTION;
    }
    if (state[SDL_SCANCODE_PAGEUP]){
        keys &= SCREEN_TOP_RIGHT;
    }
    if (state[SDL_SCANCODE_PAGEDOWN]){
        keys &= SCREEN_BOTTOM_RIGHT;
    }
    if (state[SDL_SCANCODE_HOME]){
        keys &= SCREEN_TOP_LEFT;
    }
    if (state[SDL_SCANCODE_END]){
        keys &= SCREEN_BOTTOM_LEFT;
    }
    if (state[SDL_SCANCODE_M]){
        keys &= MUTE;
    }

    if (this->controls != keys){
        this->cpu->ptInterrupt->Trigger();
    }

    this->controls = keys;

    this->cpu->mmu->StoreShort(0, this->controls);
}
